### PR TITLE
For #26957 - Exit search dialog when interacting with home fragment

### LINKED ISSUE #26957: [Bug]: The Tab pickup CFR cannot be dismissed while the search bar is active
### Prerequisites: 
On Desktop or another mobile device visit some webpages making sure you are signed in with the same account.

### Steps to reproduce
1. Perform a clear data and launch the Firefox application.
2. From the "Pick up where you left" section tap on Sign in button and sign in with the same account.
3. Dismiss the Onboarding.
4. While the search bar is active, scroll up.
5. Once the Synced tabs section is displayed, tap outside the "Your tabs are syncing! Pick up where you left off on your other device" message or tap the "X" button from the CFR. 
6. Observe the behavior.

### Expected behaviour

The Tab pickup CFR is dismissed when tapping outside the prompt or when tapping the "X" button.

### Actual behaviour

The Tab pickup CFR cannot be dismissed by tapping outside the prompt or when tapping the "X" button while the search bar is active.

### Device name

OnePlus 9 Pro  

### Android version

Android 12

### Firefox release type

Firefox Nightly

### Firefox version

Firefox Nightly 106.0a1 - 09/13

### Device logs

_No response_

### Additional information

Please note this issue is also reproducible with Poco M4 Pro (Android 12), Xiaomi Pad5 (Android 11) and other devices.

https://user-images.githubusercontent.com/89388888/190157865-de640b18-2efa-47ec-a8b6-04b04e6e70bd.mp4

https://user-images.githubusercontent.com/108326843/189911366-58f7fe10-502a-410e-9d67-876be85113dc.mp4

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21553)

### PR DESCRIPTION
**Pull request by Alexandru2909**
Dismiss the search dialog when interacting with the home fragment. The code handling clicks from `SearchDialogFragment` for homescreen items has been removed.

All the buttons from the homescreen can be interacted with:

https://user-images.githubusercontent.com/35462038/201940422-5981f9be-42f6-4be4-a2f1-c898607ef56e.mp4

After the first click on the CFR, the search dialog is dismissed and the user can interact with the it.

https://user-images.githubusercontent.com/35462038/201940973-b4f846f0-8c46-4b4b-9233-f9e51c0c640f.mp4

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #26957

### CODE REVIEW

File: app/src/test/java/org/mozilla/fenix/home/pocket/DefaultPocketStoriesControllerTest.kt [hunk 1]
```diff
@@ -277,79 +274,41 @@ class DefaultPocketStoriesControllerTest {
     }
 
     @Test
-    fun `WHEN a story is clicked THEN search is dismissed and then its link opened`() {
+    fun `WHEN a story is clicked THEN its link opened`() {
```
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/home/pocket/DefaultPocketStoriesControllerTest.kt:277]:
```suggestion
    fun `WHEN a story is clicked THEN its link is opened`() {
```

File: app/src/test/java/org/mozilla/fenix/home/pocket/DefaultPocketStoriesControllerTest.kt [hunk 2]
```diff
@@ -277,79 +274,41 @@ class DefaultPocketStoriesControllerTest {
...
 
         verifyOrder {
-            navController.navigateUp()
             homeActivity.openToBrowserAndLoad(story.url, true, BrowserDirection.FromHome)
         }
     }
 
     @Test
-    fun `WHEN discover more is clicked THEN search is dismissed and then its link opened`() {
+    fun `WHEN discover more is clicked THEN its link opened`() {
```
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/home/pocket/DefaultPocketStoriesControllerTest.kt:290]:
```suggestion
    fun `WHEN discover more is clicked THEN its link is opened`() {
```

**Review by gabrielluong — APPROVED** [review]
Please consider doing a squash landing here. I don't believe there's any benefit for landing 11 commits for this fix and having that reflected in the history especially when someone else is trying to understand the fix while reviewing the git history. I do like the reverts as a nice touch in terms of a code review, but at the same time, I also ended up just looking at all the files change altogether during the review. 

To be clear, I am not saying we shouldn't do multi-commits where it makes sense to do smaller single purpose commits. However, I am looking at this PR and thinking if we did land 11 commits where we are mostly removing code and only 5 lines in the `SearchDialogFragment` contains the core of the fix, it would seem a bit overkill.

**Comment by gabrielluong** [comment]
@Alexandru2909 Before landing this, could you also check in with @kbrosnan about the drawback with the keyboard being dismissed after clicking on the private mode button? Since this is causing a change in behaviour that you described, I am gonna add a do-not-land label since this could be unexpected to product. I am wondering if we could retrigger showing the keyboard if we clicked on the private mode button to maintain the current behaviour.

**Comment by gabrielluong** [comment]
> @Alexandru2909 Before landing this, could you also check in with @kbrosnan about the drawback with the keyboard being dismissed after clicking on the private mode button? Since this is causing a change in behaviour that you described, I am gonna add a do-not-land label since this could be unexpected to product. I am wondering if we could retrigger showing the keyboard if we clicked on the private mode button to maintain the current behaviour.

@Alexandru2909 I saw the review request, but I wanted to know what was the resolution to this before I did the review.

**Comment by Alexandru2909** [comment]
> > @Alexandru2909 Before landing this, could you also check in with @kbrosnan about the drawback with the keyboard being dismissed after clicking on the private mode button? Since this is causing a change in behaviour that you described, I am gonna add a do-not-land label since this could be unexpected to product. I am wondering if we could retrigger showing the keyboard if we clicked on the private mode button to maintain the current behaviour.
> 
> @Alexandru2909 I saw the review request, but I wanted to know what was the resolution to this before I proceed. My concern prior was that this was introducing an unexpected behaviour (keyboard being dismissed after clicking on the private mode button) and this might have unintended consequences from a Product point of view (eg, we are making it harder to start searching or entering a URL after entering private mode). In my mind, the PR would be landable if we were able to maintain the existing behaviour with entering private mode.

Hi @gabrielluong ! in the last force-push I added the `HomeFragmentLayout` class which decides based on the clicked element from the homescreen if we should dismiss the search dialog. I haven't got any response from UX/product and with help from @Mugurell I worked on a method to maintain the current behavior for the PB button. This is the reason I re-requested a review.

**Comment by gabrielluong** [comment]
> Hi @gabrielluong ! in the last force-push I added the `HomeFragmentLayout` class which decides based on the clicked element from the homescreen if we should dismiss the search dialog. I haven't got any response from UX/product and with help from @Mugurell I worked on a method to maintain the current behavior for the PB button. This is the reason I re-requested a review.

Okay great, just wanted to know that the current behaviour was maintained in your new push.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragmentLayout.kt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import android.util.AttributeSet
+import android.view.MotionEvent
+import android.view.MotionEvent.ACTION_UP
+import androidx.coordinatorlayout.widget.CoordinatorLayout
+import androidx.navigation.findNavController
+import mozilla.components.support.ktx.android.view.findViewInHierarchy
+import org.mozilla.fenix.R
+
+/**
+ * Parent layout for [HomeFragment], used to dismiss the search dialog when interacted with.
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragmentLayout.kt:18]:
```suggestion
 * Parent layout for [HomeFragment] used to dismiss the [SearchDialogFragment] when interacting with elements in the [HomeFragment].
```
This was my attempt to try to make this comment better.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragmentLayout.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ */
+class HomeFragmentLayout @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0,
+) : CoordinatorLayout(context, attrs, defStyleAttr) {
+
+    /**
+     * Checks if the motion event touches the private browsing button.
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragmentLayout.kt:27]:
```suggestion
     * Returns whether or not the motion event is touching the private browsing button.
```

File: app/src/main/java/org/mozilla/fenix/home/HomeFragmentLayout.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        val nav = findNavController()
+        // If the private button is touched from the searchDialogFragment, then it should not be
+        // dismissed, to allow the user to continue the search.
+        if (ev?.action == ACTION_UP &&
+            nav.currentDestination?.id == R.id.searchDialogFragment &&
+            !isTouchingPrivateButton(ev.x, ev.y)
+        ) {
+            nav.popBackStack()
+        }
+        return super.onInterceptTouchEvent(ev)
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragmentLayout.kt:51]:
```suggestion
        val nav = findNavController()

        // If the private button is touched from the [SearchDialogFragment], then it should not be
        // dismissed to allow the user to continue the search.
        if (ev?.action == ACTION_UP &&
            nav.currentDestination?.id == R.id.searchDialogFragment &&
            !isTouchingPrivateButton(ev.x, ev.y)
        ) {
            nav.popBackStack()
        }
        
        return super.onInterceptTouchEvent(ev)
```
Just correcting the comma usage here and added some new lines to help separate out the code to improve readability.

**Review by gabrielluong — APPROVED** [review]

**Comment by gabrielluong** [comment]
Depending on how confident you're about this PR, I would be okay with landing this before the code freeze this Thursday. This would be a nice fix to get in. Otherwise, let's wait until after the merge date. Let's also consider squashing the merge.

**Pull request merged**